src/common/Thread.h

```diff
@@ -44,6 +52,57 @@ class Thread {
   Thread();
   virtual ~Thread();
 
+  // !!!IMPORTANT!!!
+  // for correct behavior, thread ids need to be removed from this map
+  // on thread exit or else old thread names might be returned if the 
+  // pthread_t (which is actually a pointer) is reused for a different
+  // thread
+  // this infra is to help Log::dump_recent() print the correct thread
+  // names just before exit which otherwise causes pthread_getname_np()
+  // to SEGFAULT due to a bad/non-existant pthread_t id, probably
+  // because the thread ceases to exist by the time pthread_getname_np()
+  // is called
+  static void save_thread_name(unsigned int global_id, std::string& name) {
+    Thread::threadid_name_spinlock.lock();
+
+    ceph_assert(Thread::threadid_name.find(global_id) == Thread::threadid_name.end());
+    
+    /* save the global thread id in the TLS
+     * we do not allocate extra memory to store the 32bit global ID
+     * we just type cast the 32bit value to a (void*) and save it in the TLS
+     */
+    pthread_once(&global_thread_id_key_once, global_thread_id_key_alloc);
```

Comment:
C++ has the `thread_local` that takes from you the responsibility for this low level stuff.